share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -240,9 +249,60 @@ public void startup(
 
         log.info("Starting up.");
         numPartitions = shareGroupTopicPartitionCount.getAsInt();
+        setupRecordPruning();
         log.info("Startup complete.");
     }
 
+    // visibility for tests
+    void setupRecordPruning() {
+        timer.add(new TimerTask(config.shareCoordinatorTopicPruneIntervalMs()) {
+            @Override
+            public void run() {
+                runtime.activeTopicPartitions().forEach(tp -> {
+                    performRecordPruning(tp);
+                });
+                // perpetual recursion
+                setupRecordPruning();
+            }
+        });
+    }
+
+    // visibility for tests
+    void performRecordPruning(TopicPartition tp) {
+        runtime.scheduleWriteOperation(
+            "write-state-record-prune",
+            tp,
+            Duration.ofMillis(config.shareCoordinatorWriteTimeoutMs()),
+            ShareCoordinatorShard::lastRedundantOffset
+        ).whenComplete((result, exception) -> {
+            if (exception != null) {
+                Errors error = Errors.forException(exception);
+                // These errors might result from partition metadata not loaded
+                // or shard re-election. Will cause unnecessary noise, hence not logging
+                if (!(error.equals(Errors.COORDINATOR_LOAD_IN_PROGRESS) || error.equals(Errors.NOT_COORDINATOR))) {
+                    log.error("Last redundant offset lookup threw an error.", exception);
+                }
+                return;
+            }
+            result.ifPresent(
+                off -> {
+                    // guard and optimization
+                    if (off == Long.MAX_VALUE || off <= 0) {
+                        log.warn("Last redundant offset value {} not suitable to make delete call for {}.", off, tp);
+                        return;
+                    }
+
+                    log.info("Pruning records in {} till offset {}.", tp, off);
+                    try {
+                        writer.deleteRecords(tp, off);
+                    } catch (Exception e) {
+                        log.error("Failed to delete records in {} till offset {}.", tp, off, e);
```

Comment:
I believe this will be printed all the time.

Reply:
will change to debug